### ISSUE TITLE
Assertion failure on `cabal build/haddock --enable-doc`

### ISSUE THREAD

**Issue by andreasabel** [haskell/cabal#8313]
Using GHC 9.2.3 and cabal master:
```
$ cabal haddock --haddock-for-hackage --enable-doc
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
Build profile: -w ghc-9.2.3 -O0
In order, the following will be built (use -v for more details):
 - OneTuple-0.3.1 (lib) (requires build)
 ...
 - generically-0.1 (lib) (requires build)
Assertion failed
CallStack (from HasCallStack):
  assert, called at src/Distribution/Client/ProjectPlanning.hs:246:5 in cabal-install-3.9.0.0-inplace:Distribution.Client.ProjectPlanning
```
Also triggered by `cabal build --enable-doc`.

It is this assertion
https://github.com/haskell/cabal/blob/23536274bb7baabe5c33140620471e897cf34cf2/cabal-install/src/Distribution/Client/ProjectPlanning.hs#L242-L248

Here is the packed up directory in which I can reproduce the bug:
[cabal-assertion-failed.zip](https://github.com/haskell/cabal/files/9173452/cabal-assertion-failed.zip)
(I hope it can be reproduced even without my `.cabal` directory...)

**Comment by Mikolaj** [comment]
Possibl related: #6659, #6006 and the tickets mentioned in the latter.

**Comment by andreasabel** [comment]
Indeed, looks like #6006.  The new thing is that even a plain `cabal haddock` fails, since it includes `--enable-documentation` by default now.

**Comment by andreasabel** [comment]
Could we have a flag `--ingore-assertions` to force `cabal` to continue?

**Comment by Mikolaj** [comment]
> Could we have a flag `--ingore-assertions` to force `cabal` to continue?

We'd need our own `assert`, but yes, should not be hard. This is in a dev build with assertions enabled, right? Not 3.8RC1?

**Comment by andreasabel** [comment]
> This is in a dev build with assertions enabled, right? Not 3.8RC1?

Yes, it is 3.9.0.0.

**Comment by ulysses4ever** [comment]
Fail to repro with the current `master` (dd312ec).

**Comment by andreasabel** [comment]
> Fail to repro with the current `master` ([dd312ec](https://github.com/haskell/cabal/commit/dd312ecc44b79945061a4cb45e15b90f26382be4)).

@ulysses4ever : I can reproduce this with latest `master` and also GHC 9.2.4.  Did you unzip my project and tried it there?

**Comment by ulysses4ever** [comment]
I tried to build OneTuple and genetically. What is "your project"?

**Comment by ulysses4ever** [comment]
Oh, sorry, now I see it

**Comment by mpickering** [comment]
I have tracked down the failure to the `setDocumentation` function in `pruneInstallPlanPass1`.

The essential issue is that `pruneInstallPlanPass1` can't modify anything in an elaborated package which affects the hash, and by disabling documentation at this point you affect the hash of the package.

**Comment by Bodigrim** [comment]
@mpickering was this issue fixed in #9446? Could it be closed now?

**Comment by andreasabel** [comment]
And if yes, in which version?  
Please apply a/the respective milestone---would be good to easily see for closed issue in which version they were resolved.

**Comment by mpickering** [comment]
Yes it's fixed.

**Comment by andreasabel** [comment]
@ulysses4ever wrote:
> @andreasabel we don't really have a system for tracking what feature appeared in which release. The easiest would be to invent and apply labels, I think, as GH milestones are pretty heavyweight imo.

We are systematically assigning issues and PRs to milestones in the Agda codebase, e.g. https://github.com/agda/agda/milestone/37?closed=1

I guess you can do the same with labels, but I thought milestone were exactly made for this...
